File: 2_CICLO/bim1/lista3/exercicio1/Exercicio1.java
import java.util.Optional;
import java.util.Scanner;

class Node {

    private int value;
    private Node next;

    public Node(int value) {
        this.value = value;
        this.next = null;
    }

    public Node(int value, Node next) {
        this.value = value;
        this.next = next;
    }

    public int value() {
        return value;
    }

    public Optional<Node> next() {
        return Optional.ofNullable(next);
    }

    public void setNext(Node next) {
        this.next = next;
    }
}

class LinkedList {

    private Node head;

    public LinkedList() {}

    public LinkedList(Node head) {
        this.head = head;
    }

    public Optional<Node> head() {
        return Optional.ofNullable(head);
    }

    public Optional<Node> get(int idx) {
        if (idx < 0) {
            return Optional.empty();
        }

        Optional<Node> node = Optional.ofNullable(head);

        for (int i = 0; i < idx && node.isPresent(); ++i) {
            node = node.get().next();
        }

        return node;
    }

    public Optional<Node> tail() {
        if (head == null) {
            return Optional.empty();
        }

        Node current = head;
        Optional<Node> next = current.next();

        while (next.isPresent()) {
            current = next.get();
            next = current.next();
        }

        return Optional.ofNullable(current);
    }

    public void clear() {
        head = null;
    }

    // - 1. Inserir número.
    public void append(Node node) {
        if (node == null) return;

        if (head().isEmpty()) {
            head = node;
            return;
        }

        Optional<Node> tail = this.tail();

        if (tail.isPresent()) {
            tail.get().setNext(node);
        }
    }

    // - 3. Pesquisar número.
    public int search(int value) {
        int pos = -1;
        Optional<Node> atual = head();

        while (atual.isPresent()) {
            Node cur = atual.get();
            pos++;

            if (cur.value() == value) {
                return pos;
            }

            atual = cur.next();
        }

        return -1;
    }
}

public class Exercicio1 {

    static final Scanner scanner = new Scanner(System.in);

    private static void pause() {
        System.out.println("[Enter] Ok");
        scanner.nextLine();
    }

    public static void clear() {
        System.out.print("\033[H\033[2J");
        System.out.flush();
    }

    private static int readInt(String message) {
        while (true) {
            try {
                System.out.print(message);
                return Integer.parseInt(scanner.nextLine());
            } catch (NumberFormatException _) {
                System.out.println("Digite um número válido.");
            }
        }
    }

    private static void fluxoInserir(LinkedList list) {
        int qtde = readInt("Digite a quantidade de números a inserir: ");

        for (int i = 0; i < qtde; ++i) {
            list.append(
                new Node(readInt("Digite o %dº número: ".formatted(i + 1)))
            );
        }

        System.out.println("Registros inseridos com sucesso!");
        pause();
    }

    private static void fluxoListar(LinkedList list) {
        Optional<Node> atual = list.head();

        if (atual.isEmpty()) {
            System.out.println("Nenhum registro inserido.");
            pause();
            return;
        }

        while (atual.isPresent()) {
            System.out.println(atual.get().value());
            atual = atual.get().next();
        }

        System.out.println("NULL");
        System.out.println();
        pause();
    }

    private static void fluxoPesquisa(LinkedList list) {
        int num = readInt("Digite o número para pesquisar: ");
        int index = list.search(num);

        if (index == -1) {
            System.out.println("Número não encontrado");
        } else {
            System.out.printf(
                "Número '%d' encontrado no índice '%d'.\n".formatted(num, index)
            );
        }

        pause();
    }

    public static void main(String[] args) {
        LinkedList list = new LinkedList();
        int choice = 0;

        do {
            clear();
            choice = readInt(
                """
                Escolha uma opção:
                - 1. Inserir número.
                - 2. Listar números.
                - 3. Pesquisar número.
                - 4. Sair.
                >\s"""
            );
            clear();

            switch (choice) {
                case 1 -> fluxoInserir(list);
                case 2 -> fluxoListar(list);
                case 3 -> fluxoPesquisa(list);
                case 4 -> {
                }
                default -> {
                    System.out.println("Posição inválida");
                    pause();
                }
            }
        } while (choice != 4);
    }
}

File: 2_CICLO/bim1/lista3/exercicio1/Exercicio1.cs
using System;

class Node
{
    private int value;
    private Node? next;

    public Node(int value)
    {
        this.value = value;
        this.next = null;
    }

    public Node(int value, Node? next)
    {
        this.value = value;
        this.next = next;
    }

    public int Value()
    {
        return value;
    }

    public Node? Next()
    {
        return next;
    }

    public void SetNext(Node? next)
    {
        this.next = next;
    }
}

class LinkedList
{
    private Node? head;

    public LinkedList()
    {
    }

    public LinkedList(Node head)
    {
        this.head = head;
    }

    public Node? Head()
    {
        return head;
    }

    public Node? Get(int idx)
    {
        if (idx < 0)
        {
            return null;
        }

        Node? node = head;

        for (int i = 0; i < idx && node is not null; ++i)
        {
            node = node.Next();
        }

        return node;
    }

    public Node? Tail()
    {
        if (head is null)
        {
            return null;
        }

        Node current = head;
        Node? next = current.Next();

        while (next is not null)
        {
            current = next;
            next = current.Next();
        }

        return current;
    }

    public void Clear()
    {
        head = null;
    }

    // 1. Inserir número.
    public void Append(Node? node)
    {
        if (node is null)
        {
            return;
        }

        if (head is null)
        {
            head = node;
            return;
        }

        Node? tail = Tail();

        if (tail is not null)
        {
            tail.SetNext(node);
        }
    }

    // 3. Pesquisar número.
    public int Search(int value)
    {
        int pos = -1;
        Node? current = head;

        while (current is not null)
        {
            pos++;

            if (current.Value() == value)
            {
                return pos;
            }

            current = current.Next();
        }

        return -1;
    }
}

class Program
{
    private static void Pause()
    {
        Console.WriteLine("[Enter] Ok");
        Console.ReadLine();
    }

    private static void Clear()
    {
        Console.Clear();
    }

    private static int ReadInt(string message)
    {
        while (true)
        {
            Console.Write(message);

            if (int.TryParse(Console.ReadLine(), out int value))
            {
                return value;
            }

            Console.WriteLine("Digite um número válido.");
        }
    }

    private static void FluxoInserir(LinkedList list)
    {
        int qtde = ReadInt("Digite a quantidade de números a inserir: ");

        for (int i = 0; i < qtde; ++i)
        {
            list.Append(
                new Node(
                    ReadInt($"Digite o {i + 1}º número: ")
                )
            );
        }

        Console.WriteLine("Registros inseridos com sucesso!");
        Pause();
    }

    private static void FluxoListar(LinkedList list)
    {
        Node? current = list.Head();

        if (current is null)
        {
            Console.WriteLine("Nenhum registro inserido.");
            Pause();
            return;
        }

        while (current is not null)
        {
            Console.WriteLine(current.Value());
            current = current.Next();
        }

        Console.WriteLine("NULL");
        Console.WriteLine();
        Pause();
    }

    private static void FluxoPesquisa(LinkedList list)
    {
        int num = ReadInt("Digite o número para pesquisar: ");
        int index = list.Search(num);

        if (index == -1)
        {
            Console.WriteLine("Número não encontrado");
        }
        else
        {
            Console.WriteLine(
                $"Número '{num}' encontrado no índice '{index}'."
            );
        }

        Pause();
    }

    public static void Main()
    {
        LinkedList list = new();
        int choice;

        do
        {
            Clear();

            choice = ReadInt(
                """
                Escolha uma opção:
                - 1. Inserir número.
                - 2. Listar números.
                - 3. Pesquisar número.
                - 4. Sair.
                >
                """
            );

            Clear();

            switch (choice)
            {
                case 1:
                    FluxoInserir(list);
                    break;

                case 2:
                    FluxoListar(list);
                    break;

                case 3:
                    FluxoPesquisa(list);
                    break;

                case 4:
                    break;

                default:
                    Console.WriteLine("Posição inválida");
                    Pause();
                    break;
            }
        }
        while (choice != 4);
    }
}
